File: pm_trader/ws.py
"""Real-time CLOB WebSocket channels (threaded; websocket-client).

  Market channel -> live order book per token (``book`` snapshot + ``price_change``
    deltas) so the maker loop reacts to a mid move in ~ms instead of REST-polling,
    and the 149/s REST budget is freed for order writes + discovery.
  User channel   -> the account's real fills (``trade`` events) in real time,
    replacing REST ``get_trades`` polling.

Threaded to fit the sync engine: a reader thread per channel parses frames into
thread-safe state (a maintained book cache / a fills queue); the engine consumes
it via the same interfaces it already uses (``get_book`` / ``get_midpoint`` /
``poll_fills``). The PURE message handlers + subscribe builders are split from the
socket loop so they unit-test with no network; the connect/reconnect loop is
``pragma: no cover`` (needs a live WS). ``websocket-client`` is imported lazily
there, so this module imports fine without the dep (same pattern as ClobSubmitter).
"""

from __future__ import annotations

import json
import logging
import threading
import time
from collections import deque

from pm_trader.models import OrderBook, OrderBookLevel

MARKET_WS = "wss://ws-subscriptions-clob.polymarket.com/ws/market"
USER_WS = "wss://ws-subscriptions-clob.polymarket.com/ws/user"
_PING_INTERVAL_S = 10.0   # docs: send "PING" every 10s, server replies "PONG"

log = logging.getLogger("pm_trader.ws")


def _f(x, default=0.0):
    try:
        return float(x)
    except (TypeError, ValueError):
        return default


class MarketChannel:
    """Maintains a live order book per subscribed token from the market channel.

    Implements the engine's book-source interface (``get_book`` / ``get_midpoint``):
    returns the WS-maintained book, or ``None`` / ``0.0`` when a token has no fresh
    snapshot yet (just subscribed) so the caller falls back to REST.
    """

    def __init__(self, url: str = MARKET_WS) -> None:
        self.url = url
        self._lock = threading.Lock()
        # token -> {"bids": {price: size}, "asks": {price: size}}
        self._levels: dict[str, dict[str, dict[float, float]]] = {}
        self._ts: dict[str, float] = {}           # token -> last update (monotonic)
        self._tokens: set[str] = set()            # desired subscription set
        self._on_price = None                     # callback(token, mid) on updates
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None
        self._ws = None                           # live connection (set in the loop)

    def set_price_callback(self, fn) -> None:
        """Register ``fn(token_id, mid)``, fired (off-lock) on each mid update for a
        subscribed token. Used by the runner's reflex to cancel on a fast move."""
        self._on_price = fn

    # -- pure message handling (unit-tested) --------------------------------

    @staticmethod
    def subscribe_msg(tokens) -> str:
        return json.dumps({"assets_ids": list(tokens), "type": "market",
                           "custom_feature_enabled": True})

    def handle_message(self, raw) -> None:
        """Parse one WS frame (a JSON event or array of events) into the cache.
        Non-JSON frames (e.g. ``PONG``) are ignored."""
        try:
            data = json.loads(raw)
        except (ValueError, TypeError):
            return
        touched: set[str] = set()
        for ev in (data if isinstance(data, list) else [data]):
            if not isinstance(ev, dict):
                continue
            et = ev.get("event_type")
            if et == "book":
                t = self._on_book(ev)
                if t:
                    touched.add(t)
            elif et == "price_change":
                touched.update(self._on_price_change(ev))
        # fire the price callback OUTSIDE the lock (it may do I/O / take other locks)
        cb = self._on_price
        if cb is not None:
            for t in touched:
                if t in self._tokens:
                    mid = self.get_midpoint(t)
                    if mid > 0:
                        cb(t, mid)

    def _on_book(self, ev: dict) -> str | None:
        token = ev.get("asset_id")
        if not token:
            return None
        bids = {_f(l.get("price")): _f(l.get("size")) for l in ev.get("bids") or []}
        asks = {_f(l.get("price")): _f(l.get("size")) for l in ev.get("asks") or []}
        with self._lock:
            self._levels[token] = {"bids": bids, "asks": asks}   # full snapshot
            self._ts[token] = time.monotonic()
        return token

    def _on_price_change(self, ev: dict) -> set[str]:
        changes = ev.get("price_changes") or []
        touched: set[str] = set()
        with self._lock:
            for ch in changes:
                token = ch.get("asset_id")
                if not token:
                    continue
                touched.add(token)
                book = self._levels.setdefault(token, {"bids": {}, "asks": {}})
                side = "bids" if str(ch.get("side", "")).upper() == "BUY" else "asks"
                price, size = _f(ch.get("price")), _f(ch.get("size"))
                if size <= 0:
                    book[side].pop(price, None)        # size 0 -> level removed
                else:
                    book[side][price] = size
                self._ts[token] = time.monotonic()
        return touched

    # -- engine book-source interface ---------------------------------------

    def get_book(self, token_id: str) -> OrderBook | None:
        with self._lock:
            lv = self._levels.get(token_id)
            if not lv or not lv["bids"] or not lv["asks"]:
                return None                           # no fresh two-sided snapshot
            bids = [OrderBookLevel(p, s) for p, s in lv["bids"].items() if s > 0]
            asks = [OrderBookLevel(p, s) for p, s in lv["asks"].items() if s > 0]
        if not bids or not asks:
            return None
        return OrderBook(bids=bids, asks=asks)

    def get_midpoint(self, token_id: str) -> float:
        with self._lock:
            lv = self._levels.get(token_id)
            if not lv or not lv["bids"] or not lv["asks"]:
                return 0.0
            best_bid = max(lv["bids"])
            best_ask = min(lv["asks"])
        return (best_bid + best_ask) / 2.0

    def fresh(self, token_id: str, max_age_s: float = 5.0) -> bool:
        with self._lock:
            ts = self._ts.get(token_id)
        return ts is not None and (time.monotonic() - ts) <= max_age_s

    # -- subscription management --------------------------------------------

    def set_tokens(self, tokens) -> None:
        """Update the desired subscription set; the reader (re)subscribes on connect.
        Drops cached books for tokens no longer wanted."""
        new = set(t for t in tokens if t)
        with self._lock:
            self._tokens = new
            for t in list(self._levels):        # drop any cached book no longer wanted
                if t not in new:
                    self._levels.pop(t, None)
                    self._ts.pop(t, None)
        self._resubscribe()

    def _resubscribe(self) -> None:  # pragma: no cover - needs a live socket
        ws = self._ws
        if ws is None:
            return
        with self._lock:
            tokens = list(self._tokens)
        if tokens:
            try:
                ws.send(self.subscribe_msg(tokens))
            except Exception as e:  # noqa: BLE001
                log.warning("market resubscribe failed: %s", e)

    # -- lifecycle (live socket loop; pragma no cover) ----------------------

    def start(self) -> None:  # pragma: no cover - needs a live WS
        if self._thread is not None:
            return
        self._thread = threading.Thread(target=self._run, name="ws-market", daemon=True)
        self._thread.start()

    def stop(self) -> None:  # pragma: no cover - needs a live WS
        self._stop.set()
        ws = self._ws
        if ws is not None:
            try:
                ws.close()
            except Exception:  # noqa: BLE001
                pass
        if self._thread is not None:
            self._thread.join(timeout=5.0)

    def _run(self) -> None:  # pragma: no cover - needs a live WS
        import websocket  # lazy: only the live path needs the dep
        while not self._stop.is_set():
            try:
                self._ws = websocket.create_connection(self.url, timeout=15)
                self._ws.settimeout(1.0)
                self._resubscribe()
                last_ping = time.monotonic()
                while not self._stop.is_set():
                    now = time.monotonic()
                    if now - last_ping >= _PING_INTERVAL_S:
                        self._ws.send("PING")
                        last_ping = now
                    try:
                        msg = self._ws.recv()
                    except Exception:  # timeout or transient -> loop to ping/recv
                        continue
                    if msg:
                        self.handle_message(msg)
            except Exception as e:  # noqa: BLE001 - reconnect after a backoff
                if self._stop.is_set():
                    break
                log.warning("market WS reconnecting after error: %s", e)
                self._stop.wait(2.0)
            finally:
                try:
                    if self._ws is not None:
                        self._ws.close()
                except Exception:  # noqa: BLE001
                    pass
                self._ws = None


class UserChannel:
    """Streams the account's real fills (``trade`` events) into a fills queue.

    ``poll_fills`` drains it in the same shape as ``ClobSubmitter.poll_fills`` so the
    runner uses it as a drop-in for the REST fill poll.
    """

    def __init__(self, creds: dict, condition_ids=None, *, url: str = USER_WS,
                 invert_side: bool = False) -> None:
        # creds: {"apiKey","secret","passphrase"}
        self.url = url
        self.creds = creds
        self._invert = invert_side
        self._lock = threading.Lock()
        self._queue: deque = deque()
        self._seen: set = set()                   # trade ids already enqueued (dedup)
        self._markets = set(c for c in (condition_ids or []) if c)
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None
        self._ws = None

    @staticmethod
    def subscribe_msg(creds: dict, markets) -> str:
        return json.dumps({
            "type": "user",
            "auth": {"apiKey": creds.get("apiKey") or creds.get("api_key"),
                     "secret": creds.get("secret") or creds.get("api_secret"),
                     "passphrase": creds.get("passphrase") or creds.get("api_passphrase")},
            "markets": list(markets),
        })

    def handle_message(self, raw) -> None:
        try:
            data = json.loads(raw)
        except (ValueError, TypeError):
            return
        for ev in (data if isinstance(data, list) else [data]):
            if not isinstance(ev, dict) or ev.get("event_type") != "trade":
                continue
            tid = ev.get("id")
            with self._lock:
                if tid is not None and tid in self._seen:
                    continue                       # dedup repeated status updates
                if tid is not None:
                    self._seen.add(tid)
                    if len(self._seen) > 5000:     # bound the dedup set
                        self._seen = set(list(self._seen)[-2000:])
                side = str(ev.get("side", "")).upper()
                if self._invert:
                    side = "SELL" if side == "BUY" else "BUY"
                self._queue.append({
                    "id": tid,
                    "token_id": ev.get("asset_id"),
                    "side": side,
                    "size": _f(ev.get("size")),
                    "price": _f(ev.get("price")),
                })

    def poll_fills(self) -> list[dict]:
        """Drain and return all fills received since the last call."""
        with self._lock:
            out = list(self._queue)
            self._queue.clear()
        return out

    def set_markets(self, condition_ids) -> None:
        self._markets = set(c for c in condition_ids if c)
        self._resubscribe()

    def _resubscribe(self) -> None:  # pragma: no cover - needs a live socket
        ws = self._ws
        if ws is None or not self._markets:
            return
        try:
            ws.send(self.subscribe_msg(self.creds, list(self._markets)))
        except Exception as e:  # noqa: BLE001
            log.warning("user resubscribe failed: %s", e)

    def start(self) -> None:  # pragma: no cover - needs a live WS
        if self._thread is not None:
            return
        self._thread = threading.Thread(target=self._run, name="ws-user", daemon=True)
        self._thread.start()

    def stop(self) -> None:  # pragma: no cover - needs a live WS
        self._stop.set()
        ws = self._ws
        if ws is not None:
            try:
                ws.close()
            except Exception:  # noqa: BLE001
                pass
        if self._thread is not None:
            self._thread.join(timeout=5.0)

    def _run(self) -> None:  # pragma: no cover - needs a live WS
        import websocket
        while not self._stop.is_set():
            try:
                self._ws = websocket.create_connection(self.url, timeout=15)
                self._ws.settimeout(1.0)
                self._resubscribe()
                last_ping = time.monotonic()
                while not self._stop.is_set():
                    now = time.monotonic()
                    if now - last_ping >= _PING_INTERVAL_S:
                        self._ws.send("PING")
                        last_ping = now
                    try:
                        msg = self._ws.recv()
                    except Exception:
                        continue
                    if msg:
                        self.handle_message(msg)
            except Exception as e:  # noqa: BLE001
                if self._stop.is_set():
                    break
                log.warning("user WS reconnecting after error: %s", e)
                self._stop.wait(2.0)
            finally:
                try:
                    if self._ws is not None:
                        self._ws.close()
                except Exception:  # noqa: BLE001
                    pass
                self._ws = None

File: cpp/src/pmm/ws.cpp
// src/pmm/ws.cpp — 实时 WS 频道: 纯消息处理 + book 缓存 + 订阅状态 (port of pm_trader/ws.py)
//
// 传输层 (WSS socket loop) 待补 (参考 sports-trader-cpp OpenSSL WS); start/stop 当前占位。
#include "pmm/ws.hpp"

#include <algorithm>
#include <chrono>
#include <string>
#include <thread>
#include <utility>

#include "pmm/jsonutil.hpp"
#include "pmm/net/ws_connection.hpp"

namespace pmm::ws {

namespace ju = pmm::jsonutil;
using nlohmann::json;

namespace {

double mono_now() {
    return std::chrono::duration<double>(std::chrono::steady_clock::now().time_since_epoch()).count();
}
std::string upper(std::string s) {
    for (char& c : s) c = static_cast<char>((c >= 'a' && c <= 'z') ? c - 32 : c);
    return s;
}

// wss://host/path → (host, path)。
std::pair<std::string, std::string> parse_wss(const std::string& url) {
    std::string s = url;
    const std::string pfx = "wss://";
    if (s.rfind(pfx, 0) == 0) s = s.substr(pfx.size());
    const auto slash = s.find('/');
    if (slash == std::string::npos) return {s, "/"};
    return {s.substr(0, slash), s.substr(slash)};
}

// 可中断睡眠: 期间 stop 置位则提前返回 true。
bool interruptible_sleep(std::atomic<bool>& stop, double seconds) {
    using namespace std::chrono;
    const auto deadline = steady_clock::now() + duration<double>(seconds);
    while (steady_clock::now() < deadline) {
        if (stop.load()) return true;
        std::this_thread::sleep_for(milliseconds(100));
    }
    return stop.load();
}

// 把 [{"price":..,"size":..}, ...] 解析成 price→size。
std::map<double, double> parse_levels(const json& arr) {
    std::map<double, double> out;
    if (arr.is_array()) {
        for (const auto& l : arr) {
            const json* p = ju::find(l, "price");
            const json* s = ju::find(l, "size");
            out[p ? ju::to_double(*p) : 0.0] = s ? ju::to_double(*s) : 0.0;
        }
    }
    return out;
}

}  // namespace

// ---------------------------------------------------------------------------
// MarketChannel
// ---------------------------------------------------------------------------

void MarketChannel::set_price_callback(PriceCallback fn) {
    on_price_ = std::move(fn);
}

std::string MarketChannel::subscribe_msg(const std::vector<std::string>& tokens) {
    return json{{"assets_ids", tokens}, {"type", "market"}, {"custom_feature_enabled", true}}.dump();
}

std::string MarketChannel::on_book(const json& ev) {
    const json* tk = ju::find(ev, "asset_id");
    if (tk == nullptr) return {};
    const std::string token = ju::to_str(*tk);
    if (token.empty()) return {};
    Levels lv;
    if (const json* b = ju::find(ev, "bids")) lv.bids = parse_levels(*b);
    if (const json* a = ju::find(ev, "asks")) lv.asks = parse_levels(*a);
    {
        std::lock_guard<std::mutex> g(mu_);
        levels_[token] = std::move(lv);  // 全量快照
        ts_[token] = mono_now();
    }
    return token;
}

std::set<std::string> MarketChannel::on_price_change(const json& ev) {
    std::set<std::string> touched;
    const json* changes = ju::find(ev, "price_changes");
    if (changes == nullptr || !changes->is_array()) return touched;
    std::lock_guard<std::mutex> g(mu_);
    for (const auto& ch : *changes) {
        const json* tk = ju::find(ch, "asset_id");
        if (tk == nullptr) continue;
        const std::string token = ju::to_str(*tk);
        if (token.empty()) continue;
        touched.insert(token);
        Levels& book = levels_[token];
        const std::string side = (ju::find(ch, "side") && upper(ju::to_str(*ju::find(ch, "side"))) == "BUY")
                                     ? "bids"
                                     : "asks";
        const double price = ju::find(ch, "price") ? ju::to_double(*ju::find(ch, "price")) : 0.0;
        const double size = ju::find(ch, "size") ? ju::to_double(*ju::find(ch, "size")) : 0.0;
        std::map<double, double>& m = (side == "bids") ? book.bids : book.asks;
        if (size <= 0.0) {
            m.erase(price);  // size 0 → 删档
        } else {
            m[price] = size;
        }
        ts_[token] = mono_now();
    }
    return touched;
}

void MarketChannel::handle_message(const std::string& raw) {
    last_recv_.store(mono_now());  // 任意帧 (含 PONG) = 存活
    json data;
    try {
        data = json::parse(raw);
    } catch (...) {
        return;
    }
    std::set<std::string> touched;
    const auto handle_ev = [&](const json& ev) {
        if (!ev.is_object()) return;
        const json* et = ju::find(ev, "event_type");
        const std::string type = et ? ju::to_str(*et) : "";
        if (type == "book") {
            const std::string t = on_book(ev);
            if (!t.empty()) touched.insert(t);
        } else if (type == "price_change") {
            const auto tt = on_price_change(ev);
            touched.insert(tt.begin(), tt.end());
        }
    };
    if (data.is_array()) {
        for (const auto& ev : data) handle_ev(ev);
    } else {
        handle_ev(data);
    }
    {
        std::lock_guard<std::mutex> g(mu_);
        for (const auto& t : touched) updates_[t] += 1;
    }
    // 价格回调脱锁触发 (可能做 I/O / 拿别的锁)。
    if (on_price_) {
        std::set<std::string> subs_copy;
        {
            std::lock_guard<std::mutex> g(mu_);
            subs_copy = subs_;
        }
        for (const auto& t : touched) {
            if (subs_copy.count(t) != 0) {
                const double mid = get_midpoint(t);
                if (mid > 0.0) on_price_(t, mid);
            }
        }
    }
}

std::optional<OrderBook> MarketChannel::get_book(const std::string& token_id) {
    std::lock_guard<std::mutex> g(mu_);
    auto it = levels_.find(token_id);
    if (it == levels_.end() || it->second.bids.empty() || it->second.asks.empty()) return std::nullopt;
    OrderBook ob;
    for (const auto& [p, s] : it->second.bids) {
        if (s > 0.0) ob.bids.push_back({p, s});
    }
    for (const auto& [p, s] : it->second.asks) {
        if (s > 0.0) ob.asks.push_back({p, s});
    }
    if (ob.bids.empty() || ob.asks.empty()) return std::nullopt;
    return ob;
}

double MarketChannel::get_midpoint(const std::string& token_id) {
    std::lock_guard<std::mutex> g(mu_);
    auto it = levels_.find(token_id);
    if (it == levels_.end() || it->second.bids.empty() || it->second.asks.empty()) return 0.0;
    const double best_bid = it->second.bids.rbegin()->first;  // 最高 bid
    const double best_ask = it->second.asks.begin()->first;   // 最低 ask
    return (best_bid + best_ask) / 2.0;
}

bool MarketChannel::fresh(const std::string& token_id, double max_age_s) {
    std::lock_guard<std::mutex> g(mu_);
    auto it = ts_.find(token_id);
    return it != ts_.end() && (mono_now() - it->second) <= max_age_s;
}

int MarketChannel::updates(const std::string& token_id) {
    std::lock_guard<std::mutex> g(mu_);
    auto it = updates_.find(token_id);
    return it != updates_.end() ? it->second : 0;
}

bool MarketChannel::is_live(double max_silence_s) {
    const double lr = last_recv_.load();
    return lr > 0.0 && (mono_now() - lr) <= max_silence_s;
}

void MarketChannel::apply_rest_snapshot(const std::string& token_id, const json& book) {
    if (token_id.empty() || !book.is_object()) return;
    Levels lv;
    if (const json* b = ju::find(book, "bids")) lv.bids = parse_levels(*b);
    if (const json* a = ju::find(book, "asks")) lv.asks = parse_levels(*a);
    if (lv.bids.empty() || lv.asks.empty()) return;  // 不用垃圾覆盖好 book
    std::lock_guard<std::mutex> g(mu_);
    levels_[token_id] = std::move(lv);
    ts_[token_id] = mono_now();
}

void MarketChannel::set_tokens(const std::vector<std::string>& tokens) {
    std::set<std::string> new_set;
    for (const auto& t : tokens) {
        if (!t.empty()) new_set.insert(t);
    }
    {
        std::lock_guard<std::mutex> g(mu_);
        subs_ = new_set;
        for (auto it = levels_.begin(); it != levels_.end();) {  // 丢掉不再需要的 book
            if (new_set.count(it->first) == 0) {
                ts_.erase(it->first);
                it = levels_.erase(it);
            } else {
                ++it;
            }
        }
    }
    // TODO(transport): resubscribe 当连接活着。
}

std::vector<std::string> MarketChannel::tokens() {
    std::lock_guard<std::mutex> g(mu_);
    return {subs_.begin(), subs_.end()};
}

void MarketChannel::start() {
    if (reader_.joinable()) return;
    stop_.store(false);
    reader_ = std::thread([this] { run(); });
}

void MarketChannel::stop() {
    stop_.store(true);
    {
        std::lock_guard<std::mutex> lk(conn_mu_);
        if (active_conn_ != nullptr) active_conn_->shutdown_socket();  // 立即中断阻塞的 recv
    }
    if (reader_.joinable()) reader_.join();
}

MarketChannel::~MarketChannel() { stop(); }

void MarketChannel::run() {
    const auto [host, path] = parse_wss(MARKET_WS);
    while (!stop_.load()) {
        std::vector<std::string> toks = tokens();
        if (toks.empty()) {  // 无订阅 → 不持空闲连接
            if (interruptible_sleep(stop_, 0.5)) break;
            continue;
        }
        pmm::net::WsConnection conn(host, path, 2000);
        if (!conn.connect()) {
            if (interruptible_sleep(stop_, 2.0)) break;  // 重连退避
            continue;
        }
        {
            std::lock_guard<std::mutex> lk(conn_mu_);
            active_conn_ = &conn;
        }
        std::vector<std::string> subscribed = toks;
        (void)conn.send_text(subscribe_msg(subscribed));
        double last_ping = mono_now();
        while (!stop_.load()) {
            std::vector<std::string> cur = tokens();
            if (cur.empty()) break;  // 全退订 → 断连
            if (cur != subscribed) {  // 订阅集变化 → 重订
                (void)conn.send_text(subscribe_msg(cur));
                subscribed = cur;
            }
            const double now = mono_now();
            if (now - last_ping >= PING_INTERVAL_S) {
                (void)conn.send_text("PING");  // 文本 PING (PM 回文本 PONG)
                last_ping = now;
            }
            const pmm::net::WsMessage msg = conn.recv();
            if (msg.kind == pmm::net::WsMessage::Closed) break;
            if (msg.kind == pmm::net::WsMessage::Text && !msg.text.empty()) handle_message(msg.text);
        }
        {
            std::lock_guard<std::mutex> lk(conn_mu_);
            active_conn_ = nullptr;
        }
        conn.close();
        if (!stop_.load() && interruptible_sleep(stop_, 2.0)) break;
    }
}

// ---------------------------------------------------------------------------
// UserChannel
// ---------------------------------------------------------------------------

UserChannel::UserChannel(Creds creds, bool invert_side)
    : creds_(std::move(creds)), invert_(invert_side) {}

std::string UserChannel::subscribe_msg(const Creds& creds, const std::vector<std::string>& markets) {
    return json{{"type", "user"},
                {"auth",
                 {{"apiKey", creds.api_key}, {"secret", creds.secret}, {"passphrase", creds.passphrase}}},
                {"markets", markets}}
        .dump();
}

void UserChannel::handle_message(const std::string& raw) {
    json data;
    try {
        data = json::parse(raw);
    } catch (...) {
        return;
    }
    const auto handle_ev = [&](const json& ev) {
        if (!ev.is_object()) return;
        const json* et = ju::find(ev, "event_type");
        if (et == nullptr || ju::to_str(*et) != "trade") return;
        const json* idp = ju::find(ev, "id");
        const std::string tid = idp ? ju::to_str(*idp) : "";
        std::lock_guard<std::mutex> g(mu_);
        if (!tid.empty()) {
            if (seen_.count(tid) != 0) return;  // 去重
            seen_.insert(tid);
            seen_fifo_.push_back(tid);
            if (seen_fifo_.size() > 5000) {  // 真 FIFO: 淘汰最旧, 绝不丢刚插入的 (防重复成交被放行)
                seen_.erase(seen_fifo_.front());
                seen_fifo_.pop_front();
            }
        }
        std::string side = upper(ju::find(ev, "side") ? ju::to_str(*ju::find(ev, "side")) : "");
        if (invert_) side = (side == "BUY") ? "SELL" : "BUY";
        json f;
        f["id"] = tid;
        f["token_id"] = ju::find(ev, "asset_id") ? ju::to_str(*ju::find(ev, "asset_id")) : "";
        f["side"] = side;
        f["size"] = ju::find(ev, "size") ? ju::to_double(*ju::find(ev, "size")) : 0.0;
        f["price"] = ju::find(ev, "price") ? ju::to_double(*ju::find(ev, "price")) : 0.0;
        queue_.push_back(std::move(f));
    };
    if (data.is_array()) {
        for (const auto& ev : data) handle_ev(ev);
    } else {
        handle_ev(data);
    }
}

std::vector<json> UserChannel::poll_fills() {
    std::lock_guard<std::mutex> g(mu_);
    std::vector<json> out(queue_.begin(), queue_.end());
    queue_.clear();
    return out;
}

void UserChannel::set_markets(const std::vector<std::string>& condition_ids) {
    std::set<std::string> m;
    for (const auto& c : condition_ids) {
        if (!c.empty()) m.insert(c);
    }
    {
        std::lock_guard<std::mutex> g(mu_);
        markets_ = std::move(m);
    }
    // TODO(transport): resubscribe。
}

std::vector<std::string> UserChannel::markets() {
    std::lock_guard<std::mutex> g(mu_);
    return {markets_.begin(), markets_.end()};
}

void UserChannel::start() {
    if (reader_.joinable()) return;
    stop_.store(false);
    reader_ = std::thread([this] { run(); });
}

void UserChannel::stop() {
    stop_.store(true);
    {
        std::lock_guard<std::mutex> lk(conn_mu_);
        if (active_conn_ != nullptr) active_conn_->shutdown_socket();
    }
    if (reader_.joinable()) reader_.join();
}

UserChannel::~UserChannel() { stop(); }

void UserChannel::run() {
    const auto [host, path] = parse_wss(USER_WS);
    while (!stop_.load()) {
        std::vector<std::string> mk = markets();
        if (mk.empty()) {
            if (interruptible_sleep(stop_, 0.5)) break;
            continue;
        }
        pmm::net::WsConnection conn(host, path, 2000);
        if (!conn.connect()) {
            if (interruptible_sleep(stop_, 2.0)) break;
            continue;
        }
        {
            std::lock_guard<std::mutex> lk(conn_mu_);
            active_conn_ = &conn;
        }
        std::vector<std::string> subscribed = mk;
        (void)conn.send_text(subscribe_msg(creds_, subscribed));
        double last_ping = mono_now();
        while (!stop_.load()) {
            std::vector<std::string> cur = markets();
            if (cur.empty()) break;
            if (cur != subscribed) {
                (void)conn.send_text(subscribe_msg(creds_, cur));
                subscribed = cur;
            }
            const double now = mono_now();
            if (now - last_ping >= PING_INTERVAL_S) {
                (void)conn.send_text("PING");
                last_ping = now;
            }
            const pmm::net::WsMessage msg = conn.recv();
            if (msg.kind == pmm::net::WsMessage::Closed) break;
            if (msg.kind == pmm::net::WsMessage::Text && !msg.text.empty()) handle_message(msg.text);
        }
        {
            std::lock_guard<std::mutex> lk(conn_mu_);
            active_conn_ = nullptr;
        }
        conn.close();
        if (!stop_.load() && interruptible_sleep(stop_, 2.0)) break;
    }
}

}  // namespace pmm::ws
